selftests/unit/test_remote.py

```diff
@@ -25,154 +30,71 @@ class RemoteTestRunnerTest(unittest.TestCase):
+           basically by mocking 'Remote' and 'fabric' usage.
+
+        3) Provide a polluted JSON to be parsed by the RemoteTestRunner
+
+        4) Assert that those results are properly parsed into the
+           job's result
+        """
+        job_args = argparse.Namespace(test_result_total=1,
+                                      remote_username='username',
+                                      remote_hostname='hostname',
+                                      remote_port=22,
+                                      remote_password='password',
+                                      remote_key_file=None,
+                                      remote_timeout=60,
+                                      show_job_log=False,
+                                      mux_yaml=['~/avocado/tests/foo.yaml',
+                                                '~/avocado/tests/bar/baz.yaml'],
+                                      dry_run=True,
+                                      env_keep=None,
+                                      reference=['/tests/sleeptest.py',
+                                                 '/tests/other/test',
+                                                 'passtest.py'])
+
+        job = Job(job_args)
+        runner = avocado_runner_remote.RemoteTestRunner(job, job.result)
+        runner.check_remote_avocado = mock.Mock(return_value=(True,
+                                                              (version.MAJOR,
+                                                               version.MINOR)))
+
+        # These are mocked at their source, and will prevent fabric from
+        # trying to contact remote hosts
+        with mock.patch('avocado_runner_remote.Remote'):
+            with mock.patch('avocado_runner_remote.fabric'):
+                runner.remote = avocado_runner_remote.Remote(job_args.remote_hostname)
+
+            # This is the result that the run_suite() will get from remote.run
+            remote_run_result = process.CmdResult()
+            remote_run_result.stdout = JSON_RESULTS
+            remote_run_result.exit_status = 0
+            runner.remote.run = mock.Mock(return_value=remote_run_result)
+
+            # We have to fake the uncompressing and removal of the zip
+            # archive that was never generated on the "remote" end
+            # This test could be expand by mocking creating an actual
+            # zip file instead, but it's really overkill
+            with mock.patch('avocado_runner_remote.archive.uncompress'):
+                with mock.patch('avocado_runner_remote.os.remove'):
+                    runner.run_suite(None, None, 61)
+
+        # The job was created with dry_run so it should have a zeroed id
+        self.assertEqual(job.result.job_unique_id, '0' * 40)
+        self.assertEqual(job.result.tests_run, 1)
+        self.assertEqual(job.result.passed, 1)
+        shutil.rmtree(job.args.base_logdir)
```

Comment:
So basically this is a new test 😄. Anyway based on what you said in commit message the changes look good anyway I'll poke this later with debugger to see how well it utilizes the remote (especially I'm curious about the `RemoteTestRunner.setup()` because I remember it used not to be called in some versions of Avocado)

Reply:
Btw this also suffers the `tmpdir` issue (on failure it's not cleaned)